waku/v2/waku_core/topics/sharding.nim

```diff
@@ -0,0 +1,113 @@
+## Waku autosharding utils
+##
+## See 51/WAKU2-RELAY-SHARDING RFC: https://rfc.vac.dev/spec/51/#automatic-sharding
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  nimcrypto,
+  std/math,
+  std/sequtils,
+  std/algorithm,
+  std/strutils,
+  stew/endians2,
+  stew/results,
+  stew/byteutils
+
+import
+  ./content_topic,
+  ./pubsub_topic
+
+const ClusterIndex* = 49152
+const GenerationZeroShardsCount* = 5
+
+type ShardsPriority = seq[tuple[topic: NsPubsubTopic, value: float64]]
+
+proc applyWeight(hashValue: uint64, weight: float64): float64 =
+  -weight / math.ln(float64(hashValue) / float64(high(uint64)))
+
+proc hashOrder*(x, y: (NsPubsubTopic, float64)): int =
+    cmp(x[1], y[1])
+
+proc weightedShardList*(topic: NsContentTopic, shardCount: int, weights: seq[float64]): Result[ShardsPriority, string] =
+  ## Returns the ordered list of shards and their priority values.
+
+  if weights.len != shardCount:
+    return err("Must provide weights for every shards")
+
+  let shardsNWeights = zip(toSeq(0..shardCount), weights)
+
+  var list = newSeq[(NsPubsubTopic, float64)](shardCount)
+
+  for (shard, weight) in shardsNWeights:
+    let pubsub = NsPubsubTopic.staticSharding(ClusterIndex, uint16(shard))
+    let bytes = toBytes($topic) & toBytes($pubsub)
+    let hash = sha256.digest(bytes)
+    let hashValue = uint64.fromBytesBE(hash.data)
+    let value = applyWeight(hashValue, weight)
```

Comment:
We would shard apps not content topics? Totally possible but it would be worst at spreading the messages evenly on shards.

Maybe we could suggest topics in a format like /name/routing/version/topic/codec
egg. defluencer/main/v1/hashtags/cbor

Routing would be app specific and we would hash both name and routing to select a shard.
egg.
defluencer/main/v1/hashtags/cbor -> shard #1
defluencer/main/v1/likes/cbor -> shard #1
defluencer/main/v1/share/cbor -> shard #1
defluencer/aux/v1/comment/cbor -> shard #2

Reply:
I would definitely include some "namespacing" or "grouping" in the content topic format - e.g. Kubernetes resources use `group`, `version` and `kind` (GVK) to ensure this

```
apiVersion: core/v1
kind: pod
...

apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
...
```
we could define the content-topic with these fields:

```
/group/version/kind/...
```

* group represents the app namespace (e.g. status.im, waku.org...)
* version allows apps to update the protocols without breaking existing code/implementations
* kind allows apps to specify/understand what is the topic for

Similar to what @SionoiS has above, just giving names already used in other technologies

Reply:
> We would shard apps not content topics?

Well, yes and no. In general the most bandwidth-efficient routing for an app would be to have all its content topics routed on the same shard (because underlying relay would only need single shard subscription), but perhaps that is too coarse-grained for proper message distribution as you point out. This is why I suggest also including the first few bytes of the `name` portion. If that complicates things a lot, perhaps we can harden the gen 0 concept then only on hashing the app portion? It's not great distribution, but it's better than nothing and has the advantage of being very simple.

`/group/version/kind/`

Agree, though this is mostly analogous to what we intend with existing format recommendation of `/app/version/name` with added `/encoding` suffix.

Reply:
I'm trying to think of the disadvantages of adding prefixes? Parsing would be more complicated but that's manageable.

Adding another implicit prefix for routing purpose could do the trick. Hash the default routing prefix + app's name.

By default we would shard apps but then using the prefixes would allow splitting an app's traffic undo another shard.

I kinda see a way to have the same topic on multiple shards, if the prefixes were not included in the filtering of messages.

/0/unbiased/main+aux/myapp/1/mytopic/cbor The + sign would indicate that this topic is using both shards?
Do we even need this feature?

Reply:
I think in this scenario the Status Communities example is a good one.

There are two scenarios:
1. Messages of the same community should be on the same shard
3. Chat messages, media messages and control messages of the same community should be segregated between shards.

It will be difficult to have a perfect balance in the first version of auto-sharding. All we can do is pave the way to auto-balancing (that's a thing we are aiming down the line, right? out-of-scope tho).

Looking at `/application/version/name/codec`.

The only rule on content topic is that it must be empty or valid so if an app uses: `foobar` then I guess the whole content topics should be used.

Status does not even use this content topics format at this stage and would probably need to change it.

I think it's good enough to take `/application/version` to hash:

- One user is unlikely to be interested at different `version`s at a given time. If there's a new version, then a migration is in place. The user might look at both topic for a limited period but that's it. Hence, letting auto-sharding segregate on `version` is fine.

- Any app with <10k users can just use the content topic _as is_/_as designed_. All messages have a content topic starting with `/application/version` and will end up on the same shard.

- Larger apps, such as Status Communities, are comparable to "several apps". Each community is an app in itself in some ways. with possibly different network topology (members share resources vs owner pays operator). In this case, then said app, being aware of the hashing protocol, should seed `application` appropriately:
    - `/status-<4 bytes of community pubkey hash>/0/<name>/proto`
    - For larger communities, meaning we consciously segregate messages types: `/status-<4 bytes of community pubkey hash>-ctrl/0/<name>/proto`

Finally, in terms of content topic based protocols (filter, store), the application can fine tune the data downloaded by light client by varying the `name` part of the content topic.

With this proposal, we keep things simple by default, while enabling more fine tune control for larger apps.

Reply:
So instead of the application's name, the meaning of ```application``` would be what applications this topic has.

So I guess no need for the third prefix. This does the same but more elegantly.

Maybe rename that part ```purpose``` instead of ```application```?

Reply:
The downside with ` purpose`. Is that it's not obvious what it means. One more thing that developers need to grasp. This opens the risk of developers using multiple purpose values for no good reason and hence putting traffic on several shards.

I think it's easier to not have to explain to 100s of developers that ` purpose``  Must remain consistent across topic and only have to explain to larger projects (few far in between) how to scale up.

Reply:
True!